File: ruby/lib/rayz/ray.rb
module Rayz
  class Ray
    attr_reader :origin, :direction, :time

    def initialize(origin:, direction:, time: 0.0)
      @origin = origin
      @direction = direction
      @time = time  # For motion blur
    end

    def position(t)
      @origin + @direction * t
    end

    def transform(matrix)
      new_origin = Rayz::Util.matrix_multiplied_by_tuple(matrix, @origin)
      new_direction = Rayz::Util.matrix_multiplied_by_tuple(matrix, @direction)
      Ray.new(origin: new_origin, direction: new_direction, time: @time)
    end
  end
end

File: python/rayz/ray.py
from __future__ import annotations


class Ray:
    def __init__(self, origin, direction, time: float = 0.0) -> None:
        self.origin = origin
        self.direction = direction
        self.time = time

    def position(self, t: float):
        return self.origin + self.direction * t

    def transform(self, matrix) -> Ray:
        return Ray(matrix * self.origin, matrix * self.direction, time=self.time)
